### PR TITLE
Promiscuous interfaces

### PR DESCRIPTION
**Pull request by esalberg**
Add support for promiscuous interfaces.  Note: this new define also is covered in the forthcoming hiera PR.

### CODE REVIEW

**Comment by esalberg** [comment]
Working on figuring out what's wrong with my spec.

**Comment by razorsedge** [comment]
I think I need to start asking people to provide a better description in the README.md as to why someone might use these particular `defines`.  For example: is there a reason to be able to bring up a promiscuous interface with `bootproto => 'dhcp'`?

**Comment by razorsedge** [comment]
Can you put your name/details in the author and copyright section in the file header?

**Comment by esalberg** [comment]
Sorry that this got dropped.

I actually don't know if there's a reason to bring up a promiscuous interface with bootproto => dhcp (I don't work with them often - this was a request given to me separately).  Since I didn't know that it's not possible, I figured leaving it as an option wouldn't hurt.  In our implementation, we just use interface, ensure, and macaddress, and the resulting file contains DEVICE, HWADDR, PROMISC, and ONBOOT.

I can drop the bootproto part if you'd prefer and let someone else modify it later if need be.

For the readme, I kind of figured that people would know what options they wanted in the if config file and would just look to the module for that flexibility, so the module would just need to describe how, not why...

I've fixed the author.  I need to finish fixing the Travis build issue, but the details link is throwing me an error regarding loading data, so I can't see what's going on (is that broken, or is it an issue on my side?).

File: spec/defines/network_if_promisc_spec.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    end
+    it { should contain_service('network') }
+  end
+
+  context 'optional parameters' do
+    let(:title) { 'eth3' }
+    let :params do {
+      :ensure       => 'up',
+      :macaddress   => 'ef:ef:ef:ef:ef:ef',
+      :userctl      => 'yes',
```
Comment (razorsedge) [inline spec/defines/network_if_promisc_spec.rb]: Should be `:userctl      => true,`.

File: manifests/if/promisc.pp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+#     macaddress => aa:bb:cc:dd:ee:ff,
+#   }
+#
+# === Authors:
+#
+# Elyse Salberg <elyse_salberg@putnam.com>
+#
+# === Copyright:
+#
+# Copyright (C) 2011 Mike Arnold, unless otherwise noted.
```
Comment (razorsedge) [inline manifests/if/promisc.pp]: Should be `# Copyright (C) 2015 Elyse Salberg, unless otherwise noted.`.

File: templates/ifcfg-promisc.erb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+###
+### File managed by Puppet
+###
+DEVICE=<%= @interface %>
+HWADDR=<%= @macaddress %>
```
Comment (razorsedge) [inline templates/ifcfg-promisc.erb]: Should be `HWADDR=<%= @macaddy %>`.

**Comment by razorsedge** [comment]
Added notes on how to fix the tests.

**Comment by razorsedge** [comment]
One thing I am seeing, when digging into PROMISC is that /usr/share/doc/initscripts-*/sysconfig.txt is that PROMISC is "No longer supported:".  Searching through /etc/sysconfig and various bin and libexec directories on EL5 and EL6 returns no results for that string.

I am ready to press "merge", but I am wondering if this config actually works?

**Comment by esalberg** [comment]
That’s a good point – we only use this type of interface for one set of fairly legacy servers, which we’re currently replacing.  Let me check with the project owner as to whether these are still relevant after the upgrade.

And thanks for fixing my spec tests – I don’t know how I missed that. ☺

**Comment by esalberg** [comment]
I found information about needing to set this using the packet interface, which apparently can be done via adding an ifup-local script that checks to see if PROMISC is set in the ifcfg-{interface} file.

If that pans out, my recommendation would be to keep this PR but add that script as well – but I need to do some more digging / testing to make sure I’m understanding this correctly.

Thanks again!

**Comment by razorsedge** [comment]
Sounds good.

If you can, please rebase off `develop`.  It should apply cleanly.  I had to apply some fixes last night to the Gemfile to keep Travis-CI and the rspec tests happy.

**Comment by razorsedge** [comment]
@esalberg Any update on finding that script?

**Comment by esalberg** [comment]
Hi,
I actually just went out on medical leave and did not get a chance to address this before then.  I did find the information needed, but I don't know if my team will get to it before I'm back (~12 weeks).

This is the information from RedHat:
To set interface to promiscuous persistently:

1)	edit the ifcfg-ethX file and add "PROMISC=yes" to the end of the options.
2)	Create /sbin/ifup-local file (chmod +x) containing the following bash script:

RHEL5:
#!/bin/bash
RC=0

PROM=$(egrep -i 'promisc' /etc/sysconfig/network-scripts/ifcfg-$1 | awk -F"=" '{ print $2 }' | sed 's/\"//g' )

if [ "$PROM" = "yes" ];
then
    /sbin/ifconfig $1 promisc
    RC=$?
fi

exit $RC

RHEL6 (and 7?):
Change after ¿then¿:
    /sbin/ip link set $1 promisc on
    echo RC=$?

3)	Create /sbin/ifdown-local

RHEL5:
#!/bin/bash

RC=0

/sbin/ifconfig $1 -promisc
RC=$?

exit $RC
RHEL6 (and 7?):
Change command to:
/sbin/ip link set $1 promisc off

**Comment by esalberg** [comment]
I finally got this all updated - it's been a while. :)

It looks like you're not actively maintaining this module anymore - is there anything I can do to help?  I currently have four patches applied on my own fork, and it's getting a little overwhelming to manage.

**Comment by razorsedge** [comment]
I have been digging my way out of the hole of life and focusing my efforts on [razorsedge-snmp](https://forge.puppet.com/razorsedge/snmp).  This module is next on my list.

**Comment by esalberg** [comment]
Thanks so much, and I definitely understand - not trying to add any pressure. :)

**Pull request merged**